### PR TITLE
Enable tensorpipe with hip_basic backend

### PR DESCRIPTION
**Pull request by pruthvistony**
Tensorpipe branch - https://github.com/ROCmSoftwarePlatform/tensorpipe/commits/tp_rocm_54
All UTs from the file `test/distributed/rpc/cuda/test_tensorpipe_agent.py` are passing

### CODE REVIEW

**Comment by pruthvistony** [comment]
[tp_basic_pyt_test_exec_2.log](https://github.com/ROCmSoftwarePlatform/pytorch/files/9930969/tp_basic_pyt_test_exec_2.log)
[tp_pytorch_testcase.xlsx](https://github.com/ROCmSoftwarePlatform/pytorch/files/9930973/tp_pytorch_testcase.xlsx)

File: cmake/public/LoadHIP.cmake [hunk 1]
```diff
@@ -299,6 +299,8 @@ if(HIP_FOUND)
     message("ROCm version < 4.1; disablng asserts")
   endif()
 
+  enable_language(HIP)
+
```
Comment (jithunnair-amd) [inline cmake/public/LoadHIP.cmake:303]: My expectation was this this whole file would be rendered unnecessary due to this PR: https://github.com/pytorch/pytorch/pull/80469?

File: torch/csrc/distributed/rpc/tensorpipe_cuda.cpp [hunk 2]
```diff
@@ -1,14 +1,18 @@
...
 #include <c10/cuda/CUDACachingAllocator.h>
 #include <c10/cuda/CUDAGuard.h>
 #include <c10/cuda/CUDAStream.h>
 
 #include <tensorpipe/tensorpipe.h>
+#if defined(USE_ROCM)
+#include <tensorpipe/tensorpipe_hip.h>
+#else
 #include <tensorpipe/tensorpipe_cuda.h>
+#endif
```
Comment (jithunnair-amd) [inline torch/csrc/distributed/rpc/tensorpipe_cuda.cpp:15]: Why is this needed? If hipify-torch generates `tensorpipe/tensorpipe_hip.h`, it should take care of the header include hipification too?

**Review by jithunnair-amd — CHANGES_REQUESTED** [review]

File: .gitmodules [hunk 3]
```diff
@@ -151,3 +147,7 @@
 [submodule "third_party/VulkanMemoryAllocator"]
 	path = third_party/VulkanMemoryAllocator
 	url = https://github.com/GPUOpen-LibrariesAndSDKs/VulkanMemoryAllocator.git
+[submodule "third_party/tensorpipe"]
+	path = third_party/tensorpipe
+	url = https://github.com/ROCmSoftwarePlatform/tensorpipe.git
+	branch = tp_rocm_54
```
[the remark this answers is not in the document]
Comment (pruthvistony) [reply .gitmodules:153]: added 'ignore=dirty'

File: cmake/public/LoadHIP.cmake (hunk 1 above)
Comment (pruthvistony) [reply cmake/public/LoadHIP.cmake:303]: above PR [80469](https://github.com/pytorch/pytorch/pull/80469) installs LoadHIP.cmake so that it is visible to any other project trying to use it. 
Above change is to info cmake to accept HIP code. This change is NOT to related tensorpipe enablement, but a change good to have.

File: torch/csrc/distributed/rpc/tensorpipe_cuda.cpp (hunk 2 above)
Comment (pruthvistony) [reply torch/csrc/distributed/rpc/tensorpipe_cuda.cpp:15]: updated.

File: cmake/public/LoadHIP.cmake (hunk 1 above)
Comment (jithunnair-amd) [reply cmake/public/LoadHIP.cmake:303]: Does the `enable_language(HIP)` allow us to use some additional feature of HIP? Wondering why we don't need/have it in the pytorch LoadHIP.cmake.

Also, should we have a different PR to remove the LoadHIP.cmake file from tensorpipe so we can use the PyTorch file?

**Review by jithunnair-amd — APPROVED** [review]
Just need some clarifications

File: cmake/public/LoadHIP.cmake (hunk 1 above)
Comment (jithunnair-amd) [reply cmake/public/LoadHIP.cmake:303]: Sorry, I misunderstood that this is the pytorch file itself, not in tensorpipe repo. *facepalm* ..... please disregard my comments about removing this file. As discussed offline, let's just add a comment for the `enable_language(HIP)` line to provide some more context of why we added this.

File: torch/utils/hipify/cuda_to_hip_mappings.py [hunk 4]
```diff
@@ -634,6 +634,7 @@
         ("cub/device/device_reduce.cuh", ("hipcub/hipcub.hpp", CONV_INCLUDE, API_BLAS)),
         ("cub/device/device_scan.cuh", ("hipcub/hipcub.hpp", CONV_INCLUDE, API_BLAS)),
         ("nvToolsExt.h", ("roctracer/roctx.h", CONV_INCLUDE, API_ROCTX)),
+        ("tensorpipe/tensorpipe_cuda.h", ("tensorpipe/tensorpipe_hip.h", CONV_INCLUDE, API_RUNTIME)),
```
Comment (jithunnair-amd) [inline torch/utils/hipify/cuda_to_hip_mappings.py:637]: @pruthvistony As discussed, we should probably move this to a separate section for pytorch extensions, since more cases like this might come up in the future.

File: cmake/Dependencies.cmake [hunk 5]
```diff
@@ -1428,6 +1428,14 @@ if(USE_DISTRIBUTED AND USE_TENSORPIPE)
       set(TP_USE_CUDA ON CACHE BOOL "" FORCE)
       set(TP_ENABLE_CUDA_IPC ON CACHE BOOL "" FORCE)
     endif()
+    if(USE_ROCM)
+      add_compile_options(-D__HIP_PLATFORM_HCC__=1)
```
Comment (jeffdaily) [inline cmake/Dependencies.cmake:1432]: Why are we still using `__HIP_PLATFORM_HCC__`?  It is long deprecated at this point.  If this is needed for the tensorpipe code, we should also update the tensorpipe code to instead use `USE_ROCM` just like pytorch does today.

**Review by jeffdaily — APPROVED** [review]

File: torch/utils/hipify/cuda_to_hip_mappings.py (hunk 4 above)
Comment (pruthvistony) [reply torch/utils/hipify/cuda_to_hip_mappings.py:637]: Added 'API_PYT_EXT'  to identify these cases.

File: cmake/Dependencies.cmake (hunk 5 above)
Comment (pruthvistony) [reply cmake/Dependencies.cmake]: I removed `__HIP_PLATFORM_HCC__`.
For the tensorpipe there is a variable `TP_USE_ROCM` defined to separate ROCm specific code.

Comment (pruthvistony) [reply cmake/Dependencies.cmake]: HIP needed the variable - `__HIP_PLATFORM_AMD__` so added it.

9:39:16 Nov 13 03:39:16 In file included from /var/lib/jenkins/pytorch/third_party/tensorpipe/tensorpipe/common/hip_buffer.h:12,
19:39:16 Nov 13 03:39:16                  from /var/lib/jenkins/pytorch/third_party/tensorpipe/tensorpipe/common/hip_buffer.cc:10:
19:39:16 Nov 13 03:39:16 /opt/rocm/include/hip/hip_runtime.h:66:2: error: #error ("Must define exactly one of __HIP_PLATFORM_AMD__ or __HIP_PLATFORM_NVIDIA__");
19:39:16 Nov 13 03:39:16    66 | #error("Must define exactly one of __HIP_PLATFORM_AMD__ or __HIP_PLATFORM_NVIDIA__");
19:39:16 Nov 13 03:39:16       |  ^~~~~
19:39:16 Nov 13 03:39:16 In file included from /opt/rocm/include/hip/hip_runtime.h:113,

**Comment by pruthvistony** [comment]
jenkins please retest pytorch

**Comment by pruthvistony** [comment]
jenkins retest this please

**Comment by pruthvistony** [comment]
http://rocmhead.amd.com:8080/job/pytorch/job/pytorch-test-distributed-2/399/consoleText
